extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/security/HttpAuthenticator.java

```diff
@@ -277,14 +300,29 @@ public HttpCredentialTransport getCredentialTransport() {
 
     }
 
-    static class NoopCloseTask implements Runnable {
+    private static void rememberAuthAttempted(RoutingContext routingContext) {
+        routingContext.put(ATTEMPT_AUTH_INVOKED, TRUE);
+    }
 
-        static final NoopCloseTask INSTANCE = new NoopCloseTask();
+    private static boolean isAuthMechanismSelectedViaAnnotation(RoutingContext routingContext) {
+        return routingContext.get(AUTH_MECHANISM) != null;
+    }
 
-        @Override
-        public void run() {
+    private static boolean requestAlreadyAuthenticated(RoutingContext routingContext) {
+        return routingContext.get(ATTEMPT_AUTH_INVOKED) == TRUE;
+    }
 
+    static void selectAuthMechanism(RoutingContext routingContext, String authMechanism) {
+        if (requestAlreadyAuthenticated(routingContext)) {
+            throw new AuthenticationFailedException("Request has been authenticated before the '" + authMechanism
```

Comment:
Can you clarify please, what happens now when we have `@Authenticated` or `@RolesAllowed` on some method, and a matching HTTP security policy ? I'm assuming the authentication will simply be done twice, it should not cause a fault, should an `info` log message be displayed instead ?

Reply:
I have added test of these scenarios and I forgot to document it. I'll add it to docs.

> Can you clarify please, what happens now when we have @Authenticated or @RolesAllowed on some method, and a matching HTTP security policy ? 

HTTP security policy is applied before authentication mechanism is matched with request, therefore the policy is applied first and if policy attempted authentication, request fails with `AuthenticationFailedException`.

> I'm assuming the authentication will simply be done twice, it should not cause a fault, should an info log message be displayed instead ?

No, it must fail and it fails with this exact line you are pointing out (TBH before I figured this mechanism, I was very worried about introducing this feature). If you resolve identity once, it stays resolved. Just imagine that once authentication is attempted, deferred identity stays memoized and there are plenty of other places we would need to worry - `SecurityIdentityAssociation` keeps identity as well and I'm pretty sure there are other places.

Just imagine that we somehow force re-authentication with newly selected auth mechanism and same request will be authenticated for HTTP Security Policy with auth mechanism A and for RBAC with auth mechanism B. When you are selecting auth mechanism with path matching policy via configuration, you don't need to explicitly select mechanism.

Reply:
Let me put it different way - I believe once authentication has happened, it needs to be immutable for security reasons - we can't just hope there is no other place that old identity didn't stick.

**Situation is different for HTTP Permissions** because it will run before CDI request scope is activated and therefore when you augment identity and set `io.vertx.ext.web.RoutingContext#setUser`, you are still safe. After all, this permission checker runs right after authentication handler.

Reply:
Added docs note.

Reply:
Michal, would it make sense to deal with it in a dedicated PR so that we can get more people from the team to focus on this specific issue only, where the PR will only deal with blocking duplicate authentication attempts, etc. Otherwise it would be difficult for others to see what is going on, is it related to the enhancement request, or is it an independent update ?

Reply:
It is a hardening fix which would be worth a dedicated PR IMHO

Reply:
It still won't work even if I remove this insurance Sergey. This is not flaw of my design, but reality - you can't authenticate before you select authentication mechanism. Once authentication is done by HTTP permissions, you have resulting `SecurityIdentity` and you are not getting any other.

I'll give you an example:

```
quarkus.http.auth.permissions.permission1.policy=authenticated
quarkus.http.auth.permissions.permission1.path=/endpoint
```

and

```
@Path("endpoint")
public class Resource {
  
 @CodeFlow
  @GET
  public String endpoint() {
   return "endpoint";
 }

}
```

without this exception that is thrown, following request would success `curl -v http://localhost:8080/endpoint -u admin:admin`

That is not hardening, that is dangerous.

Reply:
Also, and this should probably be discussed separately, I don't understand why this matter at all. This exception makes code safe.

- you don't need to combine HTTP Security Policies with RBAC as annotation based approach at least complemented by this PR provides same capabilities.
- you still can use HTTP Security Policy non-application endpoints etc.
- we already provide features specific for disabled proactive authentication like `@Tenant`.

Reply:
@sberyozkin I'm coming with a peace offer - I can allow HTTP Security Policies to work as long as they will use the very same mechanism as is going to be selected anyway (in other words, as long as auth mechanism selection is irrelevant). How does it sound? I think I should wrote it this way at the first place, I just didn't realize there is emphasize on this, which I can see now there is.

Reply:
hey @sberyozkin - okay I thought about it and I can make it work for all HTTP Security Policies even if they use let say code-flow and later endpoint is annotated with `@Basic`. It will require more lines of code, but I can see you consider this important.

Reply:
@sberyozkin that said, I think it will be very confusing for users. at one moment you have one `SecurityIdentity` and in second moment you have different `SecurityIdentity`. The `SecurityIdentity` used by HTTP Security Policies can be augmented by the policies and later, the augmentation is going to be discarded because it belongs to the different identity. Authentication will run twice for one request.

Do you really require this?

Reply:
hello @sberyozkin ,

I made changes - now you can apply HTTP Security Policy as long as it uses same authentication mechanism.
Please see `AnnotationBasedAuthMechanismSelectionTest#testAuthenticatedHttpPolicyUsingSameMechanism` for more information.

Thank you

Reply:
Hey @michalvavrik I've missed all your recent comments, well, I was aware this PR was still open :-), but could not find the proper time to review.

So let me ask: do you agree this particular issue, how to manage HTTP Security Policy alongside annotation based authentication requirements like `@Authenticated`, `@CodeFlow`, etc, is an additional concern, and we can handle it in a follow up PR ? IMHO it would simplify the current PR significantly, and will let us focus better on this important problem better later

Reply:
> Hey @michalvavrik I've missed all your recent comments, well, I was aware this PR was still open :-), but could not find the proper time to review.

No problem, you are still the fastest reviewer in the business.

> So let me ask: do you agree this particular issue, how to manage HTTP Security Policy alongside annotation based authentication requirements like @Authenticated, @CodeFlow, etc, is an additional concern, and we can handle it in a follow up PR ?

I agree with this @sberyozkin , but it all really goes down to the line you this comment thread is pointing on (the exception). I think there are 3 possible ways from this in this PR:

1. Keep the exception. I think this is the security first option, because you still successfully can apply HTTP Security Policy, but if you had applied them on the endpoint that had selected authentication mechanism via annotation and the HTTP Security policy used different authentication mechanism, authentication will fail when the illegal state is detected. In all other cases (e.g. securing paths without `@Basic` HTTP policy is not forbidden).
2. Remove the exception. Maybe we disagree on this point, but I consider this dangerous. If you require removing the exception, we need at least to log error message that warns users something bad going on.

I'm not sure if we understand each other, I certainly don't want to keep re-iterating something that is clear to you. but I'm bit lost. We are discussing what is most likely the edge case. With the current state of this PR, this is what users can do:

1. Use annotation based auth/Z standalone
2. Combine annotation based auth/Z together with HTTP Security Policy as paths that has selected auth mechanism via annotation has policy using the same annotation.

This is what they can not do:

1. Use proactive authentication with `@Basic`
2. Select `form` with HTTP Security Policy and `basic` with annotation.

Reply:
Short version: let's propose what to do with this very line the comment thread is pointing to. I don't see other disagreement between us.

Reply:
@michalvavrik Sorry, as you see I'm definitely not the fastest reviewer :-), we all can react pretty quickly when the tie allows :-). Thanks in any case. Let me return to this PR early next week, thanks

Reply:
hello @sberyozkin, now it is possible to use HTTP Security policies, therefore closing this comment thread.